File: packages/python/peru_spatial_utils/crs/to_utm.py
import math
from typing import TypedDict

from ._params import central_meridian, derived, epsg_for
from .utm_zone_for import utm_zone_for


class UTMResult(TypedDict):
    zone: int
    hemisphere: str
    epsg: int
    easting: int
    northing: int


def to_utm(lat: float, lng: float) -> UTMResult:
    zone = utm_zone_for(lng)
    p = derived()
    a = p["a"]
    k0 = p["k0"]
    e2 = p["e2"]
    ep2 = p["ep2"]

    lat_rad = math.radians(lat)
    lon0 = math.radians(central_meridian(zone))
    delta = math.radians(lng) - lon0

    sin_lat = math.sin(lat_rad)
    cos_lat = math.cos(lat_rad)
    tan_lat = math.tan(lat_rad)

    n = a / math.sqrt(1 - e2 * sin_lat * sin_lat)
    t = tan_lat * tan_lat
    c = ep2 * cos_lat * cos_lat
    big_a = cos_lat * delta

    m = a * (
        (1 - e2 / 4 - 3 * e2 ** 2 / 64 - 5 * e2 ** 3 / 256) * lat_rad
        - (3 * e2 / 8 + 3 * e2 ** 2 / 32 + 45 * e2 ** 3 / 1024) * math.sin(2 * lat_rad)
        + (15 * e2 ** 2 / 256 + 45 * e2 ** 3 / 1024) * math.sin(4 * lat_rad)
        - (35 * e2 ** 3 / 3072) * math.sin(6 * lat_rad)
    )

    easting = (
        p["false_easting"]
        + k0
        * n
        * (
            big_a
            + (1 - t + c) * big_a ** 3 / 6
            + (5 - 18 * t + t * t + 72 * c - 58 * ep2) * big_a ** 5 / 120
        )
    )

    northing = k0 * (
        m
        + n
        * tan_lat
        * (
            big_a ** 2 / 2
            + (5 - t + 9 * c + 4 * c * c) * big_a ** 4 / 24
            + (61 - 58 * t + t * t + 600 * c - 330 * ep2) * big_a ** 6 / 720
        )
    )

    northing += p["false_northing"]

    return {
        "zone": zone,
        "hemisphere": "S",
        "epsg": epsg_for(zone),
        "easting": round(easting),
        "northing": round(northing),
    }

File: packages/go/to_utm.go
package perspatial

import "math"

func ToUTM(lat, lng float64) UtmResult {
	zone := UtmZoneFor(lng)
	central := float64(zone)*6 - 183
	for _, z := range utmCfg.Zones {
		if z.Zone == zone {
			central = z.CentralMeridian
			break
		}
	}
	epsg := 32700 + zone
	for _, z := range utmCfg.Zones {
		if z.Zone == zone {
			epsg = z.Epsg
			break
		}
	}

	e := utmCfg.Ellipsoid
	a := e.A
	f := e.F
	k0 := e.K0

	eccSq := f * (2 - f)
	ePrimeSq := eccSq / (1 - eccSq)

	latRad := lat * math.Pi / 180
	lngRad := lng * math.Pi / 180
	centralRad := central * math.Pi / 180

	N := a / math.Sqrt(1-eccSq*math.Sin(latRad)*math.Sin(latRad))
	T := math.Tan(latRad) * math.Tan(latRad)
	C := ePrimeSq * math.Cos(latRad) * math.Cos(latRad)
	A := math.Cos(latRad) * (lngRad - centralRad)

	M := a * ((1-eccSq/4-3*eccSq*eccSq/64-5*eccSq*eccSq*eccSq/256)*latRad -
		(3*eccSq/8+3*eccSq*eccSq/32+45*eccSq*eccSq*eccSq/1024)*math.Sin(2*latRad) +
		(15*eccSq*eccSq/256+45*eccSq*eccSq*eccSq/1024)*math.Sin(4*latRad) -
		(35*eccSq*eccSq*eccSq/3072)*math.Sin(6*latRad))

	easting := k0*N*(A+(1-T+C)*A*A*A/6+
		(5-18*T+T*T+72*C-58*ePrimeSq)*A*A*A*A*A/120) + e.FalseEasting

	northing := k0 * (M + N*math.Tan(latRad)*(A*A/2+
		(5-T+9*C+4*C*C)*A*A*A*A/24+
		(61-58*T+T*T+600*C-330*ePrimeSq)*A*A*A*A*A*A/720))

	northing += e.FalseNorthingSouth

	return UtmResult{
		Zone:       zone,
		Hemisphere: "S",
		Epsg:       epsg,
		Easting:    math.Round(easting),
		Northing:   math.Round(northing),
	}
}
